// @lifi/sdk/dist/cjs/services/ChainsService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("../utils/errors");
const ApiService_1 = __importDefault(require("./ApiService"));
class ChainsService {
    constructor() {
        this.chains = [];
        this.loadingPromise = this.loadAvailableChains();
    }
    async loadAvailableChains() {
        this.chains = await ApiService_1.default.getChains();
    }
    static getInstance() {
        if (!this.instance) {
            this.instance = new ChainsService();
        }
        return this.instance;
    }
    async getChainById(chainId) {
        if (this.loadingPromise) {
            await this.loadingPromise;
        }
        const chain = this.chains.find((chain) => chain.id === chainId);
        if (!chain) {
            throw new errors_1.ValidationError(`Unknown chainId passed: ${chainId}.`);
        }
        return chain;
    }
    async getChains() {
        if (this.loadingPromise) {
            await this.loadingPromise;
        }
        return this.chains;
    }
}
exports.default = ChainsService;
