// @lifi/widget/cjs/utils/navigationRoutes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.backButtonRoutes = exports.stickyHeaderRoutes = exports.navigationRoutesValues = exports.navigationRoutes = void 0;
exports.navigationRoutes = {
    home: '/',
    activeSwaps: 'active-swaps',
    bridges: 'bridges',
    exchanges: 'exchanges',
    fromChain: 'from-chain',
    fromToken: 'from-token',
    selectWallet: 'wallet',
    settings: 'settings',
    swapDetails: 'swap-details',
    swapExecution: 'swap-execution',
    swapHistory: 'swap-history',
    swapRoutes: 'swap-routes',
    toChain: 'to-chain',
    toToken: 'to-token',
    toTokenNative: 'to-token-native',
};
exports.navigationRoutesValues = Object.values(exports.navigationRoutes);
exports.stickyHeaderRoutes = [
    exports.navigationRoutes.activeSwaps,
    exports.navigationRoutes.bridges,
    exports.navigationRoutes.exchanges,
    exports.navigationRoutes.fromChain,
    exports.navigationRoutes.home,
    exports.navigationRoutes.selectWallet,
    exports.navigationRoutes.settings,
    exports.navigationRoutes.swapDetails,
    exports.navigationRoutes.swapExecution,
    exports.navigationRoutes.swapHistory,
    exports.navigationRoutes.swapRoutes,
    exports.navigationRoutes.toChain,
    exports.navigationRoutes.toTokenNative,
];
exports.backButtonRoutes = [
    exports.navigationRoutes.activeSwaps,
    exports.navigationRoutes.bridges,
    exports.navigationRoutes.exchanges,
    exports.navigationRoutes.fromChain,
    exports.navigationRoutes.fromToken,
    exports.navigationRoutes.selectWallet,
    exports.navigationRoutes.settings,
    exports.navigationRoutes.swapDetails,
    exports.navigationRoutes.swapExecution,
    exports.navigationRoutes.swapHistory,
    exports.navigationRoutes.swapRoutes,
    exports.navigationRoutes.toChain,
    exports.navigationRoutes.toToken,
    exports.navigationRoutes.toTokenNative,
];
