// @lifi/wallet-management/LiFiWalletManagement.js
import events from 'events';
import { addToActiveWallets, addToDeactivatedWallets, removeFromActiveWallets, removeFromDeactivatedWallets, } from './walletPersistance';
export class LiFiWalletManagement extends events.EventEmitter {
    constructor() {
        super(...arguments);
        this.connectedWallets = [];
        this.connect = async (wallet) => {
            try {
                await wallet.connect();
                wallet.addListener('walletAccountChanged', this.handleAccountDataChange);
                this.connectedWallets.unshift(wallet);
                removeFromDeactivatedWallets({
                    address: wallet.account?.address || '',
                    name: wallet.name,
                });
                addToActiveWallets({
                    address: wallet.account?.address || '',
                    name: wallet.name,
                });
            }
            catch (e) {
                throw e;
            }
        };
        this.disconnect = async (wallet) => {
            wallet.removeAllListeners();
            removeFromActiveWallets({
                address: wallet.account?.address || '',
                name: wallet.name,
            });
            addToDeactivatedWallets({
                address: wallet.account?.address || '',
                name: wallet.name,
            });
            wallet.disconnect();
        };
    }
    async autoConnect(wallets) {
        for (const wallet of wallets) {
            if (wallet.autoConnect) {
                await wallet.autoConnect();
                wallet.addListener('walletAccountChanged', this.handleAccountDataChange);
                this.connectedWallets.unshift(wallet);
            }
        }
    }
    handleAccountDataChange() {
        this.emit('walletChanged', this.connectedWallets);
    }
}
