// @lifi/widget/cjs/providers/SwapFormProvider/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwapFormKeyHelper = exports.SwapFormKey = void 0;
var SwapFormKey;
(function (SwapFormKey) {
    SwapFormKey["ContractOutputsToken"] = "contractOutputsToken";
    SwapFormKey["FromAmount"] = "fromAmount";
    SwapFormKey["FromChain"] = "fromChain";
    SwapFormKey["FromToken"] = "fromToken";
    SwapFormKey["ToAddress"] = "toAddress";
    SwapFormKey["ToAmount"] = "toAmount";
    SwapFormKey["ToChain"] = "toChain";
    SwapFormKey["ToContractAddress"] = "toContractAddress";
    SwapFormKey["ToContractCallData"] = "toContractCallData";
    SwapFormKey["ToContractGasLimit"] = "ToContractGasLimit";
    SwapFormKey["ToToken"] = "toToken";
    SwapFormKey["TokenSearchFilter"] = "tokenSearchFilter";
})(SwapFormKey = exports.SwapFormKey || (exports.SwapFormKey = {}));
exports.SwapFormKeyHelper = {
    getChainKey: (formType) => `${formType}Chain`,
    getTokenKey: (formType) => `${formType}Token`,
    getAmountKey: (formType) => `${formType}Amount`,
};
