// @lifi/sdk/dist/utils/multicall.js
import { Interface } from '@ethersproject/abi';
import { Contract } from '@ethersproject/contracts';
import { getRpcProvider } from '../connectors';
import { splitListIntoChunks } from './utils';
import MULTICALL_ABI from './multicallAbi.json';
const MAX_MULTICALL_SIZE = 100;
export const fetchDataUsingMulticall = async (calls, abi, chainId, multicallAddress, requireSuccess = false) => {
    // 1. create contract using multicall contract address and abi...
    const provider = await getRpcProvider(chainId);
    const multicallContract = new Contract(multicallAddress, MULTICALL_ABI, provider);
    const abiInterface = new Interface(abi);
    // split up lists into chunks to stay below multicall limit
    const chunkedList = splitListIntoChunks(calls, MAX_MULTICALL_SIZE);
    const chunkedResults = await Promise.all(chunkedList.map(async (chunkedCalls) => {
        const callData = chunkedCalls.map((call) => [
            call.address.toLowerCase(),
            abiInterface.encodeFunctionData(call.name, call.params),
        ]);
        try {
            // 3. get bytes array from multicall contract by process aggregate method...
            const { blockNumber, returnData } = await multicallContract.tryBlockAndAggregate(requireSuccess, callData);
            // 4. decode bytes array to useful data array...
            return returnData
                .map(({ success, returnData }, i) => {
                if (!success) {
                    // requested function failed
                    console.error(`Multicall unsuccessful for address "${chunkedCalls[i].address}", ` +
                        `function "${chunkedCalls[i].name}", chainId "${chainId}"`);
                    return [];
                }
                if (returnData.toString() === '0x') {
                    // requested function does probably not exist
                    console.error(`Multicall no response for address "${chunkedCalls[i].address}", ` +
                        `function "${chunkedCalls[i].name}", chainId "${chainId}"`);
                    return [];
                }
                try {
                    return abiInterface.decodeFunctionResult(chunkedCalls[i].name, returnData);
                }
                catch (e) {
                    // requested function returns other data than expected
                    console.error(`Multicall parsing unsuccessful for address "${chunkedCalls[i].address}", ` +
                        `function "${chunkedCalls[i].name}", chainId "${chainId}"`);
                    return [];
                }
            })
                .map((data) => {
                return {
                    data: data[0],
                    blockNumber: blockNumber.toNumber(),
                };
            });
        }
        catch (e) {
            // whole rpc call failed, probably an rpc issue
            console.error(`Multicall failed on chainId "${chainId}"`, chunkedList, e);
            return [];
        }
    }));
    return chunkedResults.flat();
};
